data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/ImportProcessor.java

```diff
@@ -0,0 +1,457 @@
+            .tableMetadataByTableName(params.getTableMetadataByTableName())
+            .dao(params.getDao())
+            .build();
+    ImportTaskResult importRecordResult =
+        new ImportStorageTask(taskParams, params.getDistributedStorage()).execute();
+
+    ImportTaskResult modifiedTaskResult =
+        ImportTaskResult.builder()
+            .rowNumber(importRecordResult.getRowNumber())
+            .rawRecord(importRecordResult.getRawRecord())
+            .targets(importRecordResult.getTargets())
+            .dataChunkId(dataChunk.getDataChunkId())
+            .build();
+    notifyStorageRecordCompleted(modifiedTaskResult);
+    return modifiedTaskResult;
+  }
+
+  /**
+   * Processes a complete data chunk using parallel execution. The processing mode (transactional or
+   * storage) is determined by the configured {@link ScalarDBMode}.
+   *
+   * @param dataChunk the data chunk to process
+   * @param transactionBatchSize the size of transaction batches (used only in transaction mode)
+   * @param numCores the number of CPU cores to use for parallel processing
+   * @return an {@link ImportDataChunkStatus} containing the complete processing results and metrics
+   */
+  protected ImportDataChunkStatus processDataChunk(
+      ImportDataChunk dataChunk, int transactionBatchSize, int numCores) {
+    ImportDataChunkStatus status =
+        ImportDataChunkStatus.builder()
+            .dataChunkId(dataChunk.getDataChunkId())
+            .startTime(Instant.now())
+            .status(ImportDataChunkStatusState.IN_PROGRESS)
+            .build();
+    notifyDataChunkStarted(status);
+    ImportDataChunkStatus importDataChunkStatus;
+    if (params.getScalarDBMode() == ScalarDBMode.TRANSACTION) {
+      importDataChunkStatus =
+          processDataChunkWithTransactions(dataChunk, transactionBatchSize, numCores);
+    } else {
+      importDataChunkStatus = processDataChunkWithoutTransactions(dataChunk, numCores);
+    }
+    notifyDataChunkCompleted(importDataChunkStatus);
+    return importDataChunkStatus;
+  }
+
+  /**
+   * Processes a data chunk using transaction mode with parallel batch processing. Multiple
+   * transaction batches are processed concurrently using a thread pool.
+   *
+   * @param dataChunk the data chunk to process
+   * @param transactionBatchSize the number of records per transaction batch
+   * @param numCores the maximum number of concurrent transactions to process
+   * @return an {@link ImportDataChunkStatus} containing processing results and metrics
+   */
+  private ImportDataChunkStatus processDataChunkWithTransactions(
+      ImportDataChunk dataChunk, int transactionBatchSize, int numCores) {
+    Instant startTime = Instant.now();
+    List<ImportTransactionBatch> transactionBatches =
+        splitIntoTransactionBatches(dataChunk, transactionBatchSize);
+    ExecutorService transactionBatchExecutor =
+        Executors.newFixedThreadPool(Math.min(transactionBatches.size(), numCores * 2));
```

Comment:
Could you please explain why we use `Math.min(transactionBatches.size(), numCores * 2)` here?

Reply:
@brfrn169 san,
I added this so that the number of threads created is not too much. The initial way was `numCores * 2`, I updated this considering if the transaction size provided may be small then we can create less threads. Please do let me know if you think any changes are needed here.
Thank you.

Reply:
Determining concurrency based on the number of cores is useful for CPU-bound operations, but since the import tool is I/O-bound, that may not be the best approach. Also, appropriate concurrency depends on the underlying database's latency and stability, and the number of transaction batches. Probably, it would be great if it's configurable.

Reply:
@komamitsu san,
I will add the changes to make it configurable.
Thank you.

Reply:
@komamitsu san,
I have added changes to make it configurable.
Thank you.